src/services/codefixes/inferFromUsage.ts

```diff
@@ -0,0 +1,590 @@
+
+            Debug.assert(!!references, "Found no references!");
+            Debug.assert(references.length === 1, "Found more references than expected");
+
+            return map(references[0].references, r => <Identifier>getTokenAtPosition(program.getSourceFile(r.fileName), r.textSpan.start));
+        }
+
+        function inferTypeForVariableFromUsage(token: Identifier) {
+            return InferFromReference.inferTypeFromReferences(getReferences(token), checker, cancellationToken);
+        }
+
+        function inferTypeForParameterFromUsage(containingFunction: FunctionLikeDeclaration, parameterIndex: number, isRestParameter: boolean) {
+            switch (containingFunction.kind) {
+                case SyntaxKind.Constructor:
+                case SyntaxKind.FunctionExpression:
+                case SyntaxKind.FunctionDeclaration:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.MethodSignature:
+                    const isConstructor = containingFunction.kind === SyntaxKind.Constructor;
+                    const searchToken = isConstructor ?
+                        <Token<SyntaxKind.ConstructorKeyword>>getFirstChildOfKind(containingFunction, sourceFile, SyntaxKind.ConstructorKeyword) :
+                        containingFunction.name;
+                    if (searchToken) {
+                        return InferFromReference.inferTypeForParameterFromReferences(getReferences(searchToken), parameterIndex, isConstructor, isRestParameter, checker, cancellationToken);
+                    }
+            }
+        }
+
+        function typeToString(type: Type, enclosingDeclaration: Declaration) {
+            let typeIsAccessible = true;
+
+            const writer = getSingleLineStringWriter();
+            writer.trackSymbol = (symbol, declaration, meaning) => {
+                if (checker.isSymbolAccessible(symbol, declaration, meaning, /*shouldComputeAliasToMarkVisible*/ false).accessibility !== SymbolAccessibility.Accessible) {
+                    typeIsAccessible = false;
+                }
+            };
+            writer.reportIllegalExtends = writer.reportInaccessibleThisError = () => {
+                typeIsAccessible = false;
+            };
+
+            checker.getSymbolDisplayBuilder().buildTypeDisplay(type, writer, enclosingDeclaration);
+            const typeString = typeIsAccessible ? writer.string() : undefined;
+
+            releaseStringWriter(writer);
+            writer.trackSymbol = writer.reportIllegalExtends = writer.reportInaccessibleThisError = noop;
+
+            return typeString;
+        }
+
+        function getParameterIndexInList(parameter: ParameterDeclaration, list: NodeArray<ParameterDeclaration>) {
+            for (let i = 0; i < list.length; i++) {
+                if (parameter === list[i]) return i;
+            }
+            return -1;
+        }
+
+        function getFirstChildOfKind(node: Node, sourcefile: SourceFile, kind: SyntaxKind) {

```

Comment:
There's a `findChildOfKind` which is nice and slow and allocates the lamdba you were trying to avoid here.